### PR TITLE
server: Fix potential race in CubebDeviceCollectionManager.

### PR DESCRIPTION
**Pull request by kinetiknz**
r? @ChunMinChang please

### CODE REVIEW

**Review by ChunMinChang — COMMENTED** [review]
Could you explain a bit when the race could happen? Is there a race between `CubebDeviceCollectionManager::register` and/or `CubebDeviceCollectionManager::unregister`?

File: server/src/server.rs [hunk 1]
```diff
@@ -41,61 +41,64 @@ fn error(error: cubeb::Error) -> ClientMessage {
 }
 
 struct CubebDeviceCollectionManager {
-    servers: Vec<Rc<RefCell<CubebServerCallbacks>>>,
-    devtype: cubeb::DeviceType,
+    servers: Mutex<Vec<Rc<RefCell<CubebServerCallbacks>>>>,
```
[the remark this answers is not in the document]
Comment (kinetiknz) [reply server/src/server.rs:44]: We need `RefCell` for mutable access to the `CubebServerCallbacks` held by the `Rc`.  A `Mutex` may also work, but we don't need locking there as it's managed by the added `Mutex` ensures exclusive access to the servers `Vec`.

File: server/src/server.rs [hunk 2]
```diff
@@ -41,61 +41,64 @@ fn error(error: cubeb::Error) -> ClientMessage {
...
+    registered: bool,
 }
 
 impl CubebDeviceCollectionManager {
     fn new() -> CubebDeviceCollectionManager {
         CubebDeviceCollectionManager {
-            servers: Vec::new(),
-            devtype: cubeb::DeviceType::empty(),
+            servers: Mutex::new(Vec::new()),
+            registered: false,
```
[the remark this answers is not in the document]
Comment (kinetiknz) [reply server/src/server.rs:52]: > Is it intended to use a registered variable is to register the callback outside of the critical section

That was my intent, but having considered it again while answering this I don't think it's necessary.  I'll remove `registered`.

**Comment by kinetiknz** [comment]
Repushed with `registered` removed, simplifying the logic a bit.

r? @ChunMinChang again please

**Review by ChunMinChang — APPROVED** [review]
LGTM, thanks!

**Pull request merged**